pkg/controller/traceflow/controller.go

```diff
@@ -0,0 +1,326 @@
+		if err != nil {
+			klog.Errorf("Error syncing Traceflow %s, Aborting. Error: %v", key, err)
+		}
+		// Add key to queue if retry flag is true, forget key if retry flag is false.
+		if retry {
+			c.queue.AddRateLimited(key)
+		} else {
+			c.queue.Forget(key)
+		}
+	}
+	return true
+}
+
+func (c *Controller) syncTraceflow(traceflowName string) (retry bool, err error) {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing Traceflow for %s. (%v)", traceflowName, time.Since(startTime))
+	}()
+
+	retry = false
+	tf, err := c.traceflowLister.Get(traceflowName)
+	if err != nil {
+		return
+	}
+	switch tf.Status.Phase {
+	case "", opsv1alpha1.Pending:
+		_, err = c.startTraceflow(tf)
+	case opsv1alpha1.Running:
+		retry, err = c.checkTraceflowStatus(tf)
+	default:
+		c.deallocateTag(tf)
+	}
+	return
+}
+
+func (c *Controller) startTraceflow(tf *opsv1alpha1.Traceflow) (*opsv1alpha1.Traceflow, error) {
+	// Validate if the traceflow request meets requirement.
+	if err := validate(tf); err != nil {
+		c.errorTraceflowCRD(tf, err.Error())
+		return nil, err
+	}
+
+	// Allocate data plane tag.
+	tag, err := c.allocateTag(tf)
+	if err != nil {
+		return nil, err
+	}
+	tf.Status.DataplaneTag = tag
+	tf.Status.Phase = opsv1alpha1.Running
+	payloads := fmt.Sprintf("{\"status\": {\"phase\": \"%s\", \"dataplaneTag\": %d}}",
+		tf.Status.Phase, tf.Status.DataplaneTag)
+	return c.client.OpsV1alpha1().Traceflows().Patch(context.TODO(), tf.Name, types.MergePatchType, []byte(payloads), v1.PatchOptions{})
+}
+
+func (c *Controller) checkTraceflowStatus(tf *opsv1alpha1.Traceflow) (retry bool, err error) {
+	retry = false
+	sender := false
+	receiver := false
+	for _, nodeResult := range tf.Status.Results {
+		for _, ob := range nodeResult.Observations {
+			if ob.Component == opsv1alpha1.SpoofGuard {
+				sender = true
+			}
+			if ob.Action == opsv1alpha1.Delivered || ob.Action == opsv1alpha1.Dropped {
+				receiver = true
+			}
+		}
+	}
+	if sender && receiver {
+		tf.Status.Phase = opsv1alpha1.Succeeded
+		_, err = c.client.OpsV1alpha1().Traceflows().Update(context.TODO(), tf, v1.UpdateOptions{})
+		return
+	}
+	if time.Now().UTC().Sub(tf.CreationTimestamp.UTC()).Seconds() > timeout {
+		_, err = c.errorTraceflowCRD(tf, "traceflow timeout")
+		return
+	}
+	retry = true
+	return
+}
+
+func (c *Controller) errorTraceflowCRD(tf *opsv1alpha1.Traceflow, reason string) (*opsv1alpha1.Traceflow, error) {
+	tf.Status.Phase = opsv1alpha1.Failed
+	payloads := fmt.Sprintf("{\"status\": {\"phase\": \"%s\", \"reason\": \"%s\"}}", tf.Status.Phase, reason)
+	return c.client.OpsV1alpha1().Traceflows().Patch(context.TODO(), tf.Name, types.MergePatchType, []byte(payloads), v1.PatchOptions{})
+}
+
+func validate(tf *opsv1alpha1.Traceflow) error {
+	if len(tf.Spec.Source.Namespace) == 0 || len(tf.Spec.Source.Pod) == 0 {
+		return errors.New("source pod info must be provided")
+	}
+	if len(tf.Spec.Destination.Service) != 0 {
```

Comment:
if you set preserveUnknownFields: false in validation, you will strictly allow only fields set in your CRD schema.. thus you won't need this check either

Reply:
Sure. Added TODO for validation.